### PR TITLE
Allow to set additionalProperties to a schema

### PR DESCRIPTION
**Pull request by ash2k**
**What this PR does / why we need it**:
Currently it is not possible to set `additionalProperties` in a CRD validation section of the spec to a schema. If set, validation rejects such CRD.

In json/openapi schema `additionalProperties` can be either set to a boolean true/false OR a schema object. Currently validation looks at the `Allows` field and errors out if it is `false`. But the schema may have been provided and this is an invalid behavior. It should only error out if `Allows` field is `false` and schema is not provided i.e. is `nil`.

**Release note**:
```release-note
NONE
```
/sig api-machinery
/kind bug
/area custom-resources
/cc @sttts @nikhita

### CODE REVIEW

File: staging/src/k8s.io/apiextensions-apiserver/pkg/apis/apiextensions/v1beta1/marshal.go [hunk 1]
```diff
@@ -30,7 +30,7 @@ func (s JSONSchemaPropsOrBool) MarshalJSON() ([]byte, error) {
 		return json.Marshal(s.Schema)
 	}
 
-	if s.Schema == nil && !s.Allows {
+	if !s.Allows {
```
Comment (ash2k) [inline staging/src/k8s.io/apiextensions-apiserver/pkg/apis/apiextensions/v1beta1/marshal.go]: Just cleanup. We already know here that `s.Schema == nil`.

File: staging/src/k8s.io/apiextensions-apiserver/pkg/apis/apiextensions/validation/validation_test.go [hunk 2]
```diff
@@ -458,3 +461,112 @@ func TestValidateCustomResourceDefinitionUpdate(t *testing.T) {
...
+						Plural:   "plural",
+						Singular: "singular",
+						Kind:     "kind",
+						ListKind: "listkind",
+					},
+					Scope: apiextensions.ClusterScoped,
+					Validation: &apiextensions.CustomResourceValidation{
+						OpenAPIV3Schema: &apiextensions.JSONSchemaProps{
+							AdditionalProperties: &apiextensions.JSONSchemaPropsOrBool{
+								Schema: &apiextensions.JSONSchemaProps{},
```
Comment (ash2k) [inline staging/src/k8s.io/apiextensions-apiserver/pkg/apis/apiextensions/validation/validation_test.go]: This is what I'm trying to do but it does not work.

Comment (nikhita) [reply staging/src/k8s.io/apiextensions-apiserver/pkg/apis/apiextensions/validation/validation_test.go]: While this PR makes perfect sense to me, I am a little confused if this really conforms to the Kubernetes API conventions.

When we specify a schema in `additionalProperties`, it means "the new fields are only allowed if they conform according to this schema". This makes sense in a case where you want to validate the new fields. However, this also means that you are not ignoring new fields (which goes against Kubernetes API conventions).

@sttts @ash2k wdyt?

Comment (ash2k) [reply staging/src/k8s.io/apiextensions-apiserver/pkg/apis/apiextensions/validation/validation_test.go]: I'm trying to describe restrictions for labels/annotations in the schema.

https://github.com/atlassian/smith/blob/5a992f721782c738f34201d3d74b60e0e11e221b/pkg/resources/crd_helpers.go#L77-L96

Seems like a perfectly valid use case :) If the schema author knows in advance that ALL elements in the object must conform to the spec then... this is not friendly to adding new fields later that don't. It can be expressed in a backward compatible way in the future but it cannot be made forward compatible.

**Comment by sttts** [comment]
@ash2k discussed that with @nikhita. We have originally ruled out  `additionalProperties: false` because it is an API convention in Kube to just ignore additional, but unknown fields. Having a schema instead of `true` restricts additional properties as well, similarly to `false`. Actually, the `false` case is expressible using schema, e.g. by `allOf` with contraticting sub-schemas. Hence, I see a schema restriction as a weaker form of the `false` restriction, but contradicting the API conventions the very same way.

/cc @bgrant0607

**Comment by jennybuckley** [comment]
cc @mbohlool

**Comment by ash2k** [comment]
> We have originally ruled out additionalProperties: false because it is an API convention in Kube to just ignore additional, but unknown fields.

@sttts could you please clarify what problem this restriction solves/prevents when applied to CRDs? I'm trying to understand but I don't see why this is needed.

I want to prevent my users (humans and robots) from putting rubbish into the cluster which will cause failures later. It is much better to fail earlier than later in my case. So the stricter the schema is the better.

This principle makes sense (if I understand it correctly) for types in source code because they should be forward compatible because to change them one needs to modify sources, recompile, make a release and redeploy the cluster. A lot of work indeed, multiplied by number of affected users.
For CRDs the story is different - if a forward compatibility issue is discovered, the schema can be easily updated on the CRD. Nobody is affected - no downtime required, little and trivial amount of work is needed and the risk is minimal.

To summarize: I don't see what issues this restriction solves/prevents (when applied to CRDs) but I see what issues it creates.

**Comment by sttts** [comment]
> @sttts could you please clarify what problem this restriction solves/prevents when applied to CRDs? I'm trying to understand but I don't see why this is needed.

We generally do not put constraints on unknown fields through-out the kube API. CRD should not be different than native resources. So the same rule applies as well. If we want to change that, then on the higher level of generic kube API rules.

Having said this, we actually differ from native resources because we store unknown fields. Instead we should have an option to drop fields unknown to the validation schema.

**Comment by sttts** [comment]
> Having said this, we actually differ from native resources because we store unknown fields. Instead we should have an option to drop fields unknown to the validation schema.

@deads2k @bgrant0607 ^^

**Comment by deads2k** [comment]
> We generally do not put constraints on unknown fields through-out the kube API. CRD should not be different than native resources. So the same rule applies as well. If we want to change that, then on the higher level of generic kube API rules.
> 
> Having said this, we actually differ from native resources because we store unknown fields. Instead we should have an option to drop fields unknown to the validation schema.

This is an interesting point.  Updates to newer api versions are generally safer for "normal" kube resources because name squatting on fields is impossible.  This means that the fields will be stored only if they passed admission and validation at some point.  If we do not strip the fields for CRDs, the same guarantee doesn't exist.

I think this one is worth bringing up at sig-apimachinery.  We'd need to decide whether it needs to be configurable (just how backwards compatible is backwards compatible on the beta?  This could be viewed as a mistake) and which way the default should go.

**Comment by sttts** [comment]
Added it to today's sig-api-machinery agenda.

**Comment by ash2k** [comment]
So my only option is to add a validating webhook? It will probably use a json schema 🤦‍♂️ 

For cluster administrator this is worse than just using built-in validation. The choice is between adding an extra moving piece that can fail and letting invalid data in.
For the user this is worse because webhook validation is code and not a declarative description of a valid format which may be needed. If neither are deployed then it is worse because invalid data is detected later i.e. failure feedback is delayed.

Who wins from applying this principle to CRDs? To me it looks like its a lose-lose situation. I think this principle prevents CRD developers from making forward incompatible schemas but it also prevents from accurately expressing valid constraints. I.e. cuts off valid use cases to prevent potential mistakes (which are, as I pointed out above, easy to fix). This is about finding a balance obviously. I'd give more power to the authors but documented good practices and bad practices instead of status quo.

Actually, mistakes are not prevented - one can restrict anything with a validating hook, right?

Ok, if my ramblings are ignored then at least the error message needs to be fixed - it should explicitly say that specifying anything else than `true` for `additionalProperties` is no supported. And maybe a link to this issue in validation code for those who will also think it's a bug.

**Comment by nikhita** [comment]
@ash2k Have you seen https://github.com/kubernetes/kubernetes/issues/59151? :)

> at least the error message needs to be fixed - it should explicitly say that specifying anything else than true for additionalProperties is no supported. And maybe a link to this issue in validation code for those who will also think it's a bug.

Fair enough.

**Comment by ash2k** [comment]
@nikhita yes. I hope it does not mean I cannot have a field that is a `map[string]string` equivalent in my object? The field is in the schema, the schema says it's an object. I wanted to restrict what keys are valid - trying to encode a schema for annotations and labels.

**Comment by nikhita** [comment]
> I hope it does not mean I cannot have a field that is a map[string]string equivalent in my object? 

It does **not** restrict you from having a `map[string]interface` in your object. 

https://github.com/kubernetes/kubernetes/issues/59151 means that if you are specifying a schema in the CRD, it should contain all fields (in `properties`) you want to add in the custom resource (now and in the future).

When `additionalProperties` is used, it says that "new fields need to be according to this schema" but does not specify which fields. With the above feature, you will need to explicitly list out all the fields in `properties` - which you want to use now and expect to use later - along with the schema it needs to conform.

To be clear, if any of your fields in the custom resource is a `map[string]interface`, it will consider the nested fields as well. Example:

https://github.com/atlassian/smith/blob/0c8107effa3a0cd65ed5a86d201a0f225cae5a7e/pkg/resources/crd_helpers.go#L102-L114

This is perfectly valid.

**Comment by bgrant0607** [comment]
cc @mbohlool @erictune 

I don't remember what `additionalProperties` is for, but based on a skim of the discussion:

The normal API behavior is to drop unknown fields. Fields that are not dropped are timebombs, since they effectively bypass validation until the next mutation:
https://github.com/kubernetes/kubernetes/issues/30819

Obviously, we could only drop unknown fields if we have the schema.

As for whether dropping unknown fields should be silent or not, by default it should be silent, as with built-in APIs, but I would like an option to perform server-side schema validation:
https://github.com/kubernetes/kubernetes/issues/5889

**Comment by bgrant0607** [comment]
And, I realize that we'd need to distinguish fields from maps, but we need to be able to do that anyway in order for apply to behave properly.

**Comment by ash2k** [comment]
@nikhita ok, but what about [labels and annotations](https://github.com/atlassian/smith/blob/0c8107effa3a0cd65ed5a86d201a0f225cae5a7e/pkg/resources/crd_helpers.go#L76-L95) in my schema? Are they all going to be dropped? Obviously I cannot describe all the valid annotations and labels.

> And, I realize that we'd need to distinguish fields from maps, but we need to be able to do that anyway in order for apply to behave properly.

I think this describes my concern - I have a map where I don't know the keys in advance (labels and annotations from `ObjectMeta`). And it sounds like there need to be special handling of `map` vs object fields. I of course agree that they are not the same thing semantically even though they are both an `object` in JSON representation and in JSON schema. So if they are not the same thing, can we make the "cannot specify the schema for unknown fields" rule not apply for "map" fields?

Spec for `additionalProperties` https://tools.ietf.org/html/draft-handrews-json-schema-validation-00#section-6.5.6

>  The value of "additionalProperties" MUST be a valid JSON Schema.
> 
>  This keyword determines how child instances validate for objects, and
>  does not directly validate the immediate instance itself.
>
>   Validation with "additionalProperties" applies only to the child
>   values of instance names that do not match any names in "properties",
>   and do not match any regular expression in "patternProperties".
>
>   For all such properties, validation succeeds if the child instance
>   validates against the "additionalProperties" schema.
>
>   Omitting this keyword has the same behavior as an empty schema.

**Comment by sttts** [comment]
We could use `additionalProperties: { type: "string" }` to mark those maps, maybe even universally for any schema. I.e. `additionalProperties` continues to default to true implicitly (by not being set), which will mean to drop unknown fields.

With something like that I have the suspicion that we leave the given semantics of JSON Schema with very special behaviour of `additionalProperties`.  Which other options does JSON Schema give us to mark maps as maps?

**Comment by ash2k** [comment]
I'm closing this PR in favor of an issue https://github.com/kubernetes/kubernetes/issues/59485
